### PR TITLE
fix: Fall back to SPDY when WebSocket exec is rejected for web-terminal (#848)

### PR DESCRIPTION
**Pull request by jparsai**
This PR is to cherry pick https://github.com/argoproj-labs/argocd-agent/commit/8d3a3d46906e9a09903ae68de357a2b5696ca3a3 in release-0.7.
gitope-operator PR has been merged to ensure fix is working.

## Summary by CodeRabbit

* **Bug Fixes**
  * Enhanced terminal session reliability with intelligent fallback: if WebSocket connection fails, the system automatically attempts SPDY protocol for seamless connectivity.

* **Tests**
  * Added comprehensive test coverage for terminal connection fallback behavior and edge cases.

### CODE REVIEW

**Comment by jgwest** [comment]
@coderabbitai review

**Review by jgwest — APPROVED** [review]
LGTM! 
- Fix is sufficiently contained within terminal functionality and LGTM
- It looks like E2E tests are not running on branch PRs, so I ran the E2E tests locally and they passed.
- LLM-based review did not identify any significant issues

File: agent/terminal.go [hunk 1]
```diff
@@ -344,6 +368,13 @@ func isContextCanceledError(err error) bool {
 	return strings.Contains(err.Error(), "context canceled")
 }
 
+func isWebSocketHandshakeError(err error) bool {
+	if err == nil {
+		return false
+	}
+	return strings.Contains(err.Error(), "websocket: bad handshake")
+}
```
[the remark this answers is not in the document]
Comment (jparsai) [reply agent/terminal.go:376]: Since it is backporting, I think it would be better to keep it as it is. In current version used for gorilla/websocket this is not going to break.

But yes this is a good suggestion and could be addressed main branch.
